crates/transaction-fuzzer/src/account_universe/transfer_gen.rs

```diff
@@ -60,46 +199,109 @@ impl AUTransactionGen for P2PTransferGen {
             kind,
             sender.initial_data.account.address,
             gas_object.compute_object_reference(),
-            P2P_SUCCESS_GAS_USAGE,
-            GAS_UNIT_PRICE,
+            self.gas,
+            self.gas_price,
         );
         let signed_txn = to_sender_signed_transaction(tx_data, &sender.initial_data.account.key);
-        let gas_amount = P2P_SUCCESS_GAS_USAGE * GAS_UNIT_PRICE;
-        let to_deduct = self.amount + gas_amount;
         let sender_balance = *sender.current_balances.last().unwrap();
-        // Now determine the state transition
-        let enough_max_gas = sender_balance >= gas_amount;
-        let enough_to_transfer = sender_balance >= self.amount;
-        let enough_to_succeed = sender_balance >= to_deduct;
-        let status = match (enough_max_gas, enough_to_transfer, enough_to_succeed) {
-            (true, true, true) => {
-                *sender.current_balances.last_mut().unwrap() -= to_deduct;
+        let status = match RunInfo::new(sender_balance, self) {
+            RunInfo {
+                enough_max_gas: true,
+                enough_computation_gas: true,
+                enough_to_succeed: true,
+                not_enough_gas: false,
+                gas_budget_too_high: false,
+                gas_budget_too_low: false,
+                gas_price_too_low: false,
+                gas_price_too_high: false,
+                gas_price_greater_than_budget: false,
+            } => {
+                // Fine to cast to u64 at this point, since otherwise enough_max_gas would be false
+                // since sender_balance is a u64.
+                *sender.current_balances.last_mut().unwrap() -=
+                    self.amount + p2p_success_gas(self.gas_price);
                 Ok(ExecutionStatus::Success)
             }
-            (true, _, _) => {
+            RunInfo {
+                gas_price_too_low: true,
+                ..
+            } => Err(SuiError::UserInputError {
+                error: UserInputError::GasPriceUnderRGP {
+                    gas_price: self.gas_price,
+                    reference_gas_price: 1,
+                },
+            }),
+            RunInfo {
+                gas_price_too_high: true,
```

Comment:
this is interesting, are those actually in an order that guarantees certain checks happen before others? or are we only passing one of those value being `true`?
To make it more clear, if all the values here are below are `true` (`gas_price_too_high`, `gas_budget_too_low`, etc.) what is the deal....
hope I am making sense...

Reply:
Interesting indeed. Some checks are happening before others in our `check_gas` code. We check gas price >= RGP, followed by checking gas price < max_gas_price before getting to other checks. So I suppose if all the other values are true we'll still get `GasPriceTooHigh `.

Reply:
Success and the other statuses are definitely separate, but since we want to match on everything we put that at the top.

Then there are the `UserInputErrors` and there's an ordering amongst all these due to the order in which they're checked like Emma mentioned. After that come the execution errors which are a bit more nuanced since we can get into cases around when something is insufficient gas vs you didn't have enough for the transfer and gas (in some cases you can have both! and which one you get depends on just how little gas budget you're giving it....).